# CPTAC RNA Expression pipeline
The pipeline generates the readcount per gene from GDC aligned stranded RNA-Seq BAMs.

- Pipeline URL: <https://github.com/ding-lab/cptac_rna_expression>
- Author: Liang-Bo Wang <liang-bo.wang@wustl.edu>


## Release
- v1.0 (2019-10-18): Initial release


## Processing description
Gene annotation in use is identical to GDC (`gencode.v22.annotation.gtf.gz`; md5: `291330bdcff1094bc4d5645de35e0871`), which is available on the [GDC Reference Files] page.  The readcount is generated by featureCounts (subread v1.6.4) stranded mode with parameters: `-g gene_id -t exon -Q 10 -p -B -s 2`. The readcount is later converted to FPKM and FPKM-UQ using [GDC's formula].

Note that although the gene annotation is identical to GDC, GDC readcount is done in the unstranded mode using HTSeq, so the readcounts by GDC and this pipeline are not identical.

[GDC Reference Files]: https://gdc.cancer.gov/about-data/data-harmonization-and-generation/gdc-reference-files
[GDC's formula]: https://docs.gdc.cancer.gov/Data/Bioinformatics_Pipelines/Expression_mRNA_Pipeline/#upper-quartile-fpkm


## Output format
Each sample gets its TSV in the exact same gene order. The output TSV file has the following columns:

| Column     | Description                 |
| :--------- | :-------------------------- |
| gene_id    | Ensembl gene ID             |
| read_count | read count by featureCounts |
| fpkm       | FPKM                        |
| fpkm_uq    | FPKM-UQ                     |

Use the gene information (`gencode.gene.info.v22.tsv`; md5: `0a3f1d9b0a679e2a426de36d8d74fbf9`) on the [GDC Reference Files] page to get more information about each gene.


## Installation
Create a conda environment for all the dependencies:

    conda create -n cptac_expression \
        python=3.7 \
        snakemake-minimal=5.10.0 \
        pandas=1.0.1 \
        samtools=1.9 htslib=1.9 \
        subread=1.6.4


## Pipeline execution

    # Set the batch name
    export BATCH="2020-02-25_PDA"

    cd /storage1/fs1/path/to/rna_expression_pipeline/

    # Create the batch folder
    mkdir $BATCH

    # Create samples.list. For example,
    cut -f2 CPTAC3.catalog/BamMap/compute1.BamMap.dat | tail +2 > samples.list

    # Cache the BAM map
    cp CPTAC3.catalog/BamMap/compute1.BamMap.dat compute1.BamMap.cptac3_catalog_commit_6509b69.dat

    # Create snakemake_config.json
    {
        "sample_list": "samples.list",
        "bam_map": "compute1.BamMap.cptac3_catalog_commit_6509b69.dat",
        "gdc_gtf": "../annotations/gencode.v22.annotation.gtf",
        "gdc_gene_info": "../annotations/gencode.gene.info.v22.tsv",
    }

    # Run the pipeline
    bash run.sh